### PR TITLE
add GetBlockNullifier and GetBlockRangeNullifiers #406

### LINKED ISSUE #406: RPC Method to get a stream of Nullifiers for Block Range
**What is your feature request?**
I'd like to be able to get the nullifiers first so I can know which funds are spendable for a given range

**How would this feature help you?**
It will make syncing faster and more resource efficient

fileds needed

Blockheight
for everyTx:
- txid
- sapling nullifiers
- nullifier orchard actions

### PR DESCRIPTION
**Pull request by LarryRuane**
Closes #406. These work the same as `GetBlock` and `GetBlockRange` respectively, but the compact blocks that are returned have actions containing only nullifiers; the other fields are stripped out. This reduces the size of the data stream.

### CODE REVIEW

File: frontend/service.go [hunk 1]
```diff
@@ -148,7 +148,31 @@ func (s *lwdStreamer) GetBlock(ctx context.Context, id *walletrpc.BlockID) (*wal
...
+// GetBlockNullifiers is the same as GetBlock except that it returns the compact block
+// with actions containing only the nullifiers (a subset of the full compact block).
+func (s *lwdStreamer) GetBlockNullifiers(ctx context.Context, id *walletrpc.BlockID) (*walletrpc.CompactBlock, error) {
+	if id.Height == 0 && id.Hash == nil {
+		return nil, errors.New("request for unspecified identifier")
+	}
 
+	// Precedence: a hash is more specific than a height. If we have it, use it first.
+	if id.Hash != nil {
+		// TODO: Get block by hash
```
Comment (pacu) [inline frontend/service.go:163]: Could you please track on Github with an issue?

File: frontend/service.go [hunk 2]
```diff
@@ -148,7 +148,31 @@ func (s *lwdStreamer) GetBlock(ctx context.Context, id *walletrpc.BlockID) (*wal
...
+		return nil, errors.New("gRPC GetBlock by Hash is not yet implemented")
+	}
+	cBlock, err := common.GetBlock(s.cache, int(id.Height))
+
+	if err != nil {
+		return nil, err
+	}
+	for _, tx := range cBlock.Vtx {
+		for i, action := range tx.Actions {
```
Comment (pacu) [inline frontend/service.go:172]: shouldn't sapling nullifiers be returned as well?

File: frontend/service.go [hunk 3]
```diff
@@ -177,6 +201,35 @@ func (s *lwdStreamer) GetBlockRange(span *walletrpc.BlockRange, resp walletrpc.C
...
+	for {
+		select {
+		case err := <-errChan:
+			// this will also catch context.DeadlineExceeded from the timeout
+			return err
+		case cBlock := <-blockChan:
+			err := resp.Send(cBlock)
+			for _, tx := range cBlock.Vtx {
+				for i, action := range tx.Actions {
+					tx.Actions[i] = &walletrpc.CompactOrchardAction{Nullifier: action.Nullifier}
```
Comment (pacu) [inline frontend/service.go:223]: same as above. Shouldn't Sapling nullifiers be returned as well?

**Review by pacu — CHANGES_REQUESTED** [review]
Thanks for working on this @LarryRuane 

I was wondering if we also needed nullifiers from CompactSaplingSpends to be returned as well, since DAG Sync will scan the wallet history for unspent funds that can be in the sapling pool as well.

@nuttycom @str4d I'm I thinking this correctly?

**Review by LarryRuane — COMMENTED** [review]
Thanks for the review, I need the answer to one question (see below).

File: frontend/service.go (hunk 3 above)
Comment (pacu) [reply frontend/service.go:223]: Oh cool! thanks I didn't understand it directly from the code above because I was lacking the context.

**Review by pacu — APPROVED** [review]
utACK!

This looks good to me.

@str4d @nuttycom Could you verify this is what we need for DAGsync?

File: frontend/service.go (hunk 1 above)
Comment (pacu) [reply frontend/service.go:163]: I'm not sure if we specifically need that. Do we @nuttycom ?

**Comment by LarryRuane** [comment]
@pacu - Also we need an answer to the question I asked here https://github.com/zcash/lightwalletd/pull/431#discussion_r1159313884, namely, should these new RPCs return information about Sapling outputs (i.e. `cmu`, `epk`, and `ciphertext`)? Those are returned by the existing `GetBlock` (and `GetBlockRange`), but should they be stripped out of the results of these new RPCs?

**Comment by pacu** [comment]
> @pacu - Also we need an answer to the question I asked here [#431 (comment)](https://github.com/zcash/lightwalletd/pull/431#discussion_r1159313884), namely, should these new RPCs return information about Sapling outputs (i.e. `cmu`, `epk`, and `ciphertext`)? Those are returned by the existing `GetBlock` (and `GetBlockRange`), but should they be stripped out of the results of these new RPCs?

I don't think all that is needed but I'm not the one implementing DAGsync 😅

That's probably a @str4d and @nuttycom question

**Comment by LarryRuane** [comment]
Copy-paste from slack so we have all this in one place.

@nuttycom:
>For the nullifiers, we want to strip out basically everything; we want the txid and the nullifier and I think that's it.
The idea here is that we are retrieving the minimal possible amount of data that allows us to detect when our own notes that we know about have been spent.
@str4d can you confirm?

@str4d:
>Yep. If we weren't stripping out those fields, then we'd just use `GetBlock`. For nullifiers we only need the `nf` field of spends, and the `nullifier` field of actions. And we need them to be preserved in the same relative locations they would be in `GetBlock`.

**Comment by LarryRuane** [comment]
Force-pushed an update to remove the `outputs` from Sapling transactions.

**Comment by LarryRuane** [comment]
Just to mention explicitly: This PR is compatible with darkside wallet, that is, this PR does not need to change anything with darkside wallet code. You can use darkside wallet to test these RPCs.

**Comment by LarryRuane** [comment]
@nuttycom @str4d 
The results that the new gRPCs added by this PR include the per-block information (basically the block header) that the existing `GetBlock` and `GetBlockRange` gRPCs provide as part of their results, for example:
```
  "height": "2091941",
  "hash": "3YuXLwZEKxefuXhZFrWU8ceujZcT7ldB0SNJAQAAAAA=",
  "prevHash": "qUUafE5QSka8lBVSPDRGHf88ilQvg0v2miEeAQAAAAA=",
  "time": 1684534214,
```
So first of all, I'd like to verify that this is correct. I would assume so because this isn't much data; the real point of these new gRPCs is to reduce only the rather large _per-transaction_ data except for the nullifiers.

If that's correct (we do want these new gRPCs to return the per-block information), then should these new gRPCs return the sapling and orchard tree commitment sizes that were added by PR #435? Once I have the answer to that, then I think this PR is ready to be merged.

If I had to guess, I'd say the answer is yes, because this is a small amount of data (and only per block). But if we know that it's definitely unneeded, we can leave it out.

**Comment by nuttycom** [comment]
> If I had to guess, I'd say the answer is yes, because this is a small amount of data (and only per block). But if we know that it's definitely unneeded, we can leave it out.

Hm. While I can imagine us making use of the block hash, and height (and *maybe* the prevhash and time), I can't really see a scenario where we'd be making any use of the commitment tree sizes when we're retrieving nullifier data. The reason is that these values are just referents for the commitment tree positions of the individual notes in the transaction data, but we aren't returning any of those (in this case, we're not returning any information about the outputs at all) so there really isn't any point to including the commitment tree sizes. 

At a high level, the commitment tree sizes are output-related information, and the nullifiers are input-related information, and we don't do anything relative to outputs when we're using the nullifiers to detect our notes from previous blocks being spent.

**Comment by LarryRuane** [comment]
Force pushed to rebase.

**Pull request merged**